src/nupic/ntypes/Dimensions.hpp

```diff
@@ -29,352 +29,149 @@
-   * how many dimensions they have, i.e. isOnes() returns @c true for both
-   * of them. Some linking scenarios require us to treat [1] equivalent to [1 1]
-   * etc.
-   *
-   * @param dims2
-   *        The Dimensions object being compared
-   *
-   * @returns
-   *        Whether this Dimensions object is equivalent to @a dims2.
+   * * Invalid     - Some dimension is 0.
    *
+   * There is a function to check for each of these states.
    */
-  bool operator==(const Dimensions &dims2) const;
-
-  /**
-   * The in-equivalence operator, the opposite of operator==().
-   *
-   * @param dims2
-   *        The Dimensions object being compared
-   *
-   * @returns
-   *        Whether this Dimensions object is not equivalent to @a dims2.
-   */
-  bool operator!=(const Dimensions &dims2) const;
+  bool isUnspecified() const { return(size() == 0); }
+  bool isDontcare()    const { return(size() == 1 && at(0) == 0); }
+  bool isInvalid()     const { return(getCount() == 0); }
+  bool isSpecified()   const { return(!isInvalid()); }
+  static const int DONTCARE = 0;
 
-  /**
-   *
-   * @}
-   *
-   */
+  std::string toString(bool humanReadable = true) const {
+    if (isUnspecified()) return "[unspecified]";
+    if (isDontcare())    return "[dontcare]";
+    std::stringstream ss;
+    ss << "[";
+    for (size_t i = 0; i < size(); i++) {
+      if (i)  ss << "," <<at(i);
+      else   ss << at(i);
+    }
+    ss << "] ";
+		if (humanReadable && isInvalid()) ss << "(Invalid) ";
+    return ss.str();
+  }
+
+  void save(std::ostream &f) const {
+    size_t n = size();
+    f.write((const char*)&n, sizeof(size_t));
+    if (n > 0)
+      f.write((const char*)&at(0), n * sizeof(at(0)));
+  }
+  void load(std::istream &f) {
+    size_t n;
+    f.read((char*)&n, sizeof(size_t));
+    clear();
+    if (n > 0) {
+      resize(n);
+      f.read((char*)&at(0), n * sizeof(at(0)));
+    }
+  }
 
-  friend std::ostream &operator<<(std::ostream &f, const Dimensions &d);
-  friend std::istream &operator>>(std::istream &f, Dimensions &d);
 };
+  
+
+  inline std::ostream &operator<<(std::ostream &f, const Dimensions& d) {
+    f << d.toString() << " ";
+    return f;
+  }
+  inline std::istream &operator>>(std::istream &f, Dimensions& d) { 
+    // expected format:    [val, val, val]
+    f >> std::ws;  // ignore leading whitespace
+    d.clear();
+    NTA_CHECK(f.get() == '[') << "Expecting beginning of Dimensions.";
+    if (!isdigit(f.peek())) {
+      std::string tag;
+      f >> tag;
+      if (tag == "unspecified]") {
+        // leave d empty.
+      }
+      else if (tag == "dontcare]") {
+        d.push_back(0);
+      }
+    }
+    else {
+      int c;
+      UInt32 i;
+      char buf[50];
+      while(isdigit(f.peek())) {
+        int j = 0;
+        while(isdigit(c = f.get())) {
+          buf[j++] = c;
+        }
+        buf[j] = '\0';
+        i = strtoul(buf, nullptr, 0);
+        d.push_back(i);
+        f >> std::ws;  // ignore whitespace
```

Comment:
this is a nice usecase! Thanks for showing us

Reply:
`f >> std::ws` That's a neat trick!